components/brave_wallet/browser/eth_allowance_manager.cc

```diff
@@ -0,0 +1,416 @@
+      chain_cached_data->FindString(kLastBlockNumber);
+
+  uint256_t block_number{0};
+  if (!block_number_ptr ||
+      !HexValueToUint256(*block_number_ptr, &block_number)) {
+    return;
+  }
+
+  for (const auto& ca_item : *cached_allowances_ptr) {
+    const auto* ca_dict = ca_item.GetIfDict();
+    if (!ca_dict) {
+      continue;
+    }
+    const auto* approver_address = ca_dict->FindString(kApproverAddress);
+    const auto* contract_address = ca_dict->FindString(kContractAddress);
+    const auto* spender_address = ca_dict->FindString(kSpenderAddress);
+    const auto* amount = ca_dict->FindString(kAmount);
+
+    if (!approver_address || !contract_address || !spender_address || !amount ||
+        base::ToUpperASCII(*approver_address) !=
+            base::ToUpperASCII(hex_account_address)) {
+      continue;
+    }
+
+    allowance_map.insert_or_assign(
+        GetAllowanceMapKey(*contract_address, *approver_address,
+                           *spender_address),
+        std::make_tuple(block_number,
+                        mojom::AllowanceInfo::New(chain_id, *contract_address,
+                                                  *approver_address,
+                                                  *spender_address, *amount)));
+  }
+}
+
+/**
+ * @brief Represents callback which processes data portions
+ * retrieved from logs and calls the DiscoverEthAllowancesCallback with
+ * last processed portion of data
+ *
+ * @param task_id - Id of the processing task
+ * @param chain_id - Id of the Chain, which the current data portion belongs to
+ * @param hex_account_address - Account address (whith padding), which the
+ * current data portion belongs to
+ * @param logs - chain_id and account address related Log data portion
+ * @param rawlogs - chain_id and account address related Log data portion
+ * @param error - error code
+ * @param error_message - error message
+ */
+void EthAllowanceManager::OnGetAllowances(
+    const int& task_id,
+    const std::string& chain_id,
+    const std::string& hex_account_address,
+    const std::vector<Log>& logs,
+    base::Value rawlogs,
+    mojom::ProviderError error,
+    const std::string& error_message) {
+  if (error != mojom::ProviderError::kSuccess) {
+    if (tasks.find(task_id) != tasks.end()) {
+      tasks[task_id]->MarkComplete();
+      MergeAllResultsAnCallBack();
+    }
+    return;
+  }
+
+  std::vector<Log> logs_tmp{logs};
+  std::sort(logs_tmp.begin(), logs_tmp.end(),
+            [](const Log& first, const Log& second) -> bool {
+              return std::tie(first.block_number, first.log_index) <
+                     std::tie(second.block_number, second.log_index);
+            });
+
+  // Collection of the latest allowances per contract & spender & approver.
+  std::map<std::string, std::tuple<uint256_t, mojom::AllowanceInfoPtr>>
+      allowance_map;
+  // Put cached data into the map first if available.
+  LoadCachedAllowances(chain_id, hex_account_address, allowance_map);
```

Comment:
I don't understand the whole idea of caching here.
Cache is supposed to reduce number of requests to rpc endpoint but here we are reading cache after response arrives. How is this supposed to work?

Reply:
In other words if there is no caching we still should get same results but with more requests or with more time needed.

Reply:
It reads from the preferences last known list of allowances with last checked blocks numbers to search new allowances starting from that point.